### ISSUE TITLE
Cryptographically secured PGP lambda MIME web services, where the client supplies the code

### ISSUE THREAD

**Issue by polterguy** [jstedfast/MimeKit#407]
Hi Jeff, here's an example of using some code I've been playing around with lately.

[25 minutes long video demonstrating the concept](https://www.youtube.com/watch?v=h_ycJAASZmw)

The basic idea is to use MimeKit at the core, for creating cryptographically signed PGP MME envelopes, and encrypt them, to establish trust and privacy between the client and the server, such that the client can supply the actual code to execute in the web service, allowing for basically having one web service endpoint, solving _"everything"_.

In addition to the PGP parts, this also have lots of other advantages, such as compressing the request and response (since PGP also compresses its PGP messages), and sending multiple files in one HTTP request (MIME supports _"multiparts"_), allowing for the client's code to reference files in the MIME request, and also itself decide which files to return from server (assuming it's authorized to read the files).

Basically, a secure implementation for letting the **client** supply the **code** that the web service endpoint executes ...

Thoughts ...?

**Q**; Security concerns I should think of before releasing this...?

**Comment by jstedfast** [comment]
Seems like a cool way of implementing trust for a web server that runs code uploaded by a client.

I'm probably not the best person to ask about security concerns as I'm hardly an expert, but I don't see anything obvious at least.

Anyway, thanks for sharing :)

**Comment by polterguy** [comment]
Thank you, I've had the infrastructure for evaluating code passed into the web service by the client for quite some time, but have significantly refined it to make it more secure, much thx to your baby :)

Among other things, the ws endpoint can declare a _"whitelist"_ of PGP keys, and only if the request has been cryptographically signed by one PGP key from this list, the code will be evaluated.

It also significantly reduces the bandwidth consumption, since PGP encrypting the MIME envelope also compresses it, and allows for passing in and returning attachments, etc, etc, etc. A pretty kick ass idea I'd say, and I'm quite frankly surprised nobody have thought about it before in fact, since it's such an obvious use case for MIME and PGP ... :)
